expression/builtin_compare.go

```diff
@@ -1085,21 +1086,29 @@ func tryToConvertConstantInt(ctx sessionctx.Context, isUnsigned bool, con *Const
 	}
 	dt, err = dt.ConvertTo(sc, fieldType)
 	if err != nil {
-		return con, terror.ErrorEqual(err, types.ErrOverflow)
+		if terror.ErrorEqual(err, types.ErrOverflow) {
+			return con, &Constant{
+				Value:        dt,
+				RetType:      fieldType,
+				DeferredExpr: con.DeferredExpr,
+			}, true
+		}
+		return con, nil, false
 	}
 	return &Constant{
 		Value:        dt,
 		RetType:      fieldType,
 		DeferredExpr: con.DeferredExpr,
-	}, false
+	}, nil, false
 }
 
 // RefineComparedConstant changes an non-integer constant argument to its ceiling or floor result by the given op.
-// isAlwaysFalse indicates whether the int column "con" is false.
-func RefineComparedConstant(ctx sessionctx.Context, isUnsigned bool, con *Constant, op opcode.Op) (_ *Constant, isAlwaysFalse bool) {
+// isAlways indicates whether the int column "con" is true/false.
+// The second return-val only used when isAlways is True
+func RefineComparedConstant(ctx sessionctx.Context, isUnsigned bool, con *Constant, op opcode.Op) (_ *Constant, _ *Constant, isAlways bool) {
 	dt, err := con.Eval(chunk.Row{})
 	if err != nil {
-		return con, false
+		return con, nil, false
 	}
 	sc := ctx.GetSessionVars().StmtCtx
 	intFieldType := types.NewFieldType(mysql.TypeLonglong)
```

Comment:
We can define this intFieldType as an input parameter. It should be the FiledType of the int column.

``` sql
tidb> create table t (a int);
Query OK, 0 rows affected (0.01 sec)

tidb> desc select * from t where a > 128; -- It should be TableDual for the sql as well.
+---------------------+----------+------+------------------------------------------------------------+
| id                  | count    | task | operator info                                              |
+---------------------+----------+------+------------------------------------------------------------+
| TableReader_7       | 3333.33  | root | data:Selection_6                                           |
| └─Selection_6       | 3333.33  | cop  | gt(test.t.a, 128)                                          |
|   └─TableScan_5     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+---------------------+----------+------+------------------------------------------------------------+
3 rows in set (0.00 sec)

tidb> desc select * from t where a > 11111111111111111111111111111;
+-------------+-------+------+---------------+
| id          | count | task | operator info |
+-------------+-------+------+---------------+
| TableDual_6 | 0.00  | root | rows:0        |
+-------------+-------+------+---------------+
1 row in set (0.00 sec)
```